### ISSUE TITLE
Convert to newer module specifications (.mjs)

### ISSUE THREAD

**Issue by MichaelKentBurns** [MichaelKentBurns/BibleModel#7]
Just doing some reading about Node.js related I just discovered that the original Node.js modules are now referred to as .cjs (using require and module.export syntax) is OLD and with recent ECMA script versions is superseded by .mjs (using import / export syntax) are better and supported by all recent browsers and current Node.js.   

So, I'm now determined to make this switchover as a major refactoring task.   

My plan is to write some decent jest test cases for my current classes and get them all passing.   

Then I will undertake this change.   

PS. I'm also introducing the REFACTOR tag for issues.  This will be one of the first.

**Comment by MichaelKentBurns** [comment]
https://nodejs.org/docs/latest/api/esm.html

**Comment by MichaelKentBurns** [comment]
Another great learning and reference resource is https://nodejs.org/api/documentation.html
It appears to be "the latest and greatest" as ov v23.3
It's great strength for my purposes is that (under the title in the sidebar: Modules: ECMAScript modules) the Introduction and Enabling sections CLEARLY explain the differences between the legacy module system (.js or .cjs) and the new standard (.mjs).   

The first major difference is noticed in the file extension you can have a .mjs (new module system) that is more compatible with recent browser syntax.  In particular the use of the include statement to pull in other modules, rather than the older require syntax of the early Node.js modules).    

When the difference in syntax is important there is a "CJS ESM" switch on the right margin.  When you toggle that switch you see the syntax appropriate for that module system.    For example, use the sidebar and scroll down into the 'C's and select 'Console'.    When you find the "CJS ESM" switch and toggle it, the example code changes.      This clears up a lot of the crazy sample code confusion I have seen.  Other tutorials will contain one syntax with no qualification.  If you paste that code into a modules source of the other kind, you find it throws strange errors.     

Another clarification is the different ways of doing asynchronous actions.  You have Promises, Await, and the syntax for those things is wildly divergent.   Adding confusion is the different syntax candy for handling asynchronous actions, try/catch, and error handling.   The api details make these differences clear.   I can't find a good example at the moment, but when I run across it again, I'll add another comment with the details.

**Comment by MichaelKentBurns** [comment]
Finally, I think I'm ready to switch from plain .js to .mjs files and change the import/require syntax.   I will likely run into errors along the way and discover other changes that need to be made.

As I encounter them, I'll comment here again.

**Comment by MichaelKentBurns** [comment]
I'm in the process of converting to .mjs.  I have:

1) renamed .js to .mjs files.
2) changed from old require syntax to more standard import statements.

While doing 2, I ran into the fact that the require('sqlite3') even when changed to import from 'node:sqlite3' is out of date. 

The version of node.js I installed a year ago was version v20.14.0.   The document I have been following has another great feature, and that is notes about version requirements.   
The most recent documentation is labeled "Node.js v23.3.0 documentation".

In the sidebar of that document one of the major subdivisions is labeled 'SQLite'.   That indicates that the reference should be just 'sqlite' rather than 'sqlite3'.   I always thought that having a version number in the name of a feature was strange.     

The document clearly says:   SQLite[#](https://nodejs.org/api/sqlite.html#sqlite) Added in: v22.5.0

So, this means my install of Node.js is two years out of date.    
Next, I found this document:  How to Update Node.js and NPM to the Latest Version (2024)
https://www.geeksforgeeks.org/update-node-js-and-npm-to-latest-version/#

I am going to do that and change my references from sqlite3 to simply node:sqlite .

**Comment by MichaelKentBurns** [comment]
One step I needed was to update my version of node.

Following the instructions just mentioned above, I installed the 'nvm' NodeVersionManager so I could update to the latest version of node (v23.3.0) and also to switch between different versions.
Nvm supports the install option to install a specific version.   It ends up creating a new directory in $HOME called .nvm :

michael@Logos ~ % cd .nvm
michael@Logos .nvm % ls -latr
total 0
lrwxr-xr-x    1 michael  staff    36 Nov 27 08:33 nvm.sh -> /opt/homebrew/opt/nvm/libexec/nvm.sh
lrwxr-xr-x    1 michael  staff    38 Nov 27 08:33 nvm-exec -> /opt/homebrew/opt/nvm/libexec/nvm-exec
drwxr-xr-x    3 michael  staff    96 Nov 27 08:37 .cache
drwxr-xr-x    7 michael  staff   224 Nov 27 08:37 .
drwxr-xr-x    3 michael  staff    96 Nov 27 08:37 versions
drwxr-xr-x    4 michael  staff   128 Nov 27 08:37 alias
drwxr-xr-x@ 255 michael  staff  8160 Nov 29 16:28 ..
michael@Logos .nvm % ls -l versions
total 0
drwxr-xr-x  4 michael  staff  128 Nov 29 11:03 node
michael@Logos .nvm % ls -l versions/node
total 0
drwxr-xr-x  9 michael  staff  288 Nov 29 11:03 v22.5.0
drwxr-xr-x  9 michael  staff  288 Nov 27 08:37 v23.3.0
michael@Logos .nvm % 

So, now I have v22 and v23 installed.   nvm also has a command 'use' which allows you to switch versions on the fly.  

michael@Logos ~ % node -v      
v22.5.0
michael@Logos ~ % nvm use v23
Now using node v23.3.0 (npm v10.9.0)
michael@Logos ~ % node -v
v23.3.0
michael@Logos ~ % 

Also, you can manually set your IDE to use a specific one with a specific target. 
I'm currently using WebStorm which is very similar to VS Code.  In the launch configuration for my BibleModel project I have the Node interpreter set to use ~/.nvm/versions/node/v23.3.0/bin/node

**Comment by MichaelKentBurns** [comment]
The node api documentation mentioned a couple of comments ago has told me that the newest sqlite is now imported as 'node:sqlite'.   The documentation says that this is experimental code and you need to add this option to the node command:  --experimental-sqlite 

You can to that on a command line invocation like this:

node --experimental-sqlite Bible.mjs 
Or in my  WebStorm configuration file there is a field for Node parameters: which I set to that new option.     

Thirdly, the nom (Node Package Manager) has a package.json file in the directory where your working project is.   Here is mine:

{
  "name": "biblemodel",
  "version": "1.0.0",
  "description": "database model for the Bible",
  "main": "Bible.mjs",
  "scripts": {
    "test": "jest",
    "main": "node --experimental-sqlite Bible.mjs"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/MichaelKentBurns/BibleModel.git"
  },
  "private": true,
  "dependencies": {
    "sqlite": "",
    "sqlite3": "^5.1.7"
  },
  "devDependencies": {
    "jest": "^29.7.0"
  }
}

Notice that in the scripts section I have 2 targets.  The main one (used to run this package as an application)  

With those targets, I can selectively run those target by simply using the 'rpm <script>' name (such as my  "main": "node --experimental-sqlite Bible.mjs") by simply entering the command: 

npm run main

`michael@Logos BibleModel % npm run main
(node:24853) ExperimentalWarning: CommonJS module /Users/Michael/.nvm/versions/node/v23.3.0/lib/node_modules/npm/node_modules/debug/src/node.js is loading ES Module /Users/Michael/.nvm/versions/node/v23.3.0/lib/node_modules/npm/node_modules/supports-color/index.js using require().
Support for loading ES Module in require() is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)

> biblemodel@1.0.0 main
> node --experimental-sqlite Bible.mjs

Book.mjs initializing.
Book.mjs initialized.
Bible.mjs Initializing for the first time
Bible.mjs Initializing
Bible.mjs singleton theBible created.
  ...
`

**Comment by MichaelKentBurns** [comment]
So, it is working at this point.

The biggest thing this program does at the moment is to setup a Bible object which then contains an array of Book objects, then the Bible calls Book class's loadAll function.  That queries the SQL database to load an array of all of the 66 books of the Bible.   When that is done, it also writes that array as a json file named books.json in the current directory.   As an optimization, during the load process it checks for the books.json file and if present it simply reads that in, saving a setup and execution of the sqlite.        To demonstrate that I usually delete the books.json file and run the program, then run it again to demonstrate that it reads the json file rather than invoking sqlite.  

If you inspect the two files: Notes/runs/Bible.mjs-2024-11-29-run1 and run2  you can see that only run1 includes this bit from the loadAll function:

`
Bible.mjs Requesting the loading of books...
Bible.mjs LoadAll() Loading books...
StatementSync { sourceSQL: [Getter], expandedSQL: [Getter] }
Book[1] Genesis loaded.
`
Where in run2 you see this: 
`
Bible.mjs Requesting the loading of books...
Bible.mjs LoadAll() Loading books...
A promise to load all books =  undefined
Check if books loaded without a promise to wait on.
Bible.mjs books are complete and not empty. Move to booksLoaded without waiting.
`

**Comment by MichaelKentBurns** [comment]
If you bothered to look at those two runs, you will notice that there is a lot of repetition of processing.    This is partly because of the excessive code I had to put in to wait for the asynchronous SQL process.   I had to create several states in my state machine and use the asynchronous 'promise' feature of javascript.   

One huge feature of the new version of node.js is that the node:sqlite has the feature of synchronous (ie. no waiting) processing of the SQL query.  I still have the promise hooks in there, but they are no longer necessary.      I have another major issue for that, and I think I can now take out a bunch of that code.   But, in the interesting of separating these two issues, I'm going to commit and push the changes from this issue  and then return to that other issue and clean up my code considerably.
